### PR TITLE
Add dartdocs for RenderObject

### PR DESCRIPTION
**Pull request by abarth**
... and other code in object.dart.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie

File: sky/packages/sky/lib/rendering/object.dart [hunk 1]
```diff
@@ -16,50 +16,71 @@ import 'package:vector_math/vector_math.dart';
 export 'dart:sky' show Point, Offset, Size, Rect, Color, Paint, Path;
 export 'package:sky/base/hit_test.dart' show EventDisposition, HitTestTarget, HitTestEntry, HitTestResult;
 
+/// Data associated with a [RenderObject] by its parent in the render tree
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Base class for...

File: sky/packages/sky/lib/rendering/object.dart [hunk 2]
```diff
@@ -16,50 +16,71 @@ import 'package:vector_math/vector_math.dart';
...
 }
 
+/// A place to paint
+///
+/// Rather than holding a canvas directly, render objects paint using a painting
+/// context. The painting context has a canvas, which receives the
+/// individual draw operations, and also has functions for painting child
+/// render objects.
+///
+/// When painting a child render object, the canvas held by the paintint context
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: paintint typo

File: sky/packages/sky/lib/rendering/object.dart [hunk 3]
```diff
@@ -116,6 +146,10 @@ class PaintingContext {
 
   static final Paint _disableAntialias = new Paint()..isAntiAlias = false;
 
+  /// Paint a child with a rectangular clip
+  ///
+  /// If the child needs compositing, the clip will be applied by a
+  /// compositing layer. Otherwise, the clip will be applied by the canvas.
   void paintChildWithClipRect(RenderObject child, Point childPosition, Rect clipRect) {
     // clipRect is in the parent's coordinate space
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: hoist this comment into the dartdocs.

File: sky/packages/sky/lib/rendering/object.dart [hunk 4]
```diff
@@ -132,6 +166,10 @@ class PaintingContext {
     }
   }
 
+  /// Paint a child with a rounded-rectangular clip
+  ///
+  /// If the child needs compositing, the clip will be applied by a
+  /// compositing layer. Otherwise, the clip will be applied by the canvas.
   void paintChildWithClipRRect(RenderObject child, Point childPosition, Rect bounds, sky.RRect clipRRect) {
     // clipRRect is in the parent's coordinate space
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: ditto

File: sky/packages/sky/lib/rendering/object.dart [hunk 5]
```diff
@@ -132,6 +166,10 @@ class PaintingContext {
     }
   }
 
+  /// Paint a child with a rounded-rectangular clip
+  ///
+  /// If the child needs compositing, the clip will be applied by a
+  /// compositing layer. Otherwise, the clip will be applied by the canvas.
   void paintChildWithClipRRect(RenderObject child, Point childPosition, Rect bounds, sky.RRect clipRRect) {
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: describe `bounds`

File: sky/packages/sky/lib/rendering/object.dart [hunk 6]
```diff
@@ -148,6 +186,10 @@ class PaintingContext {
     }
   }
 
+  /// Paint a child with a clip path
+  ///
+  /// If the child needs compositing, the clip will be applied by a
+  /// compositing layer. Otherwise, the clip will be applied by the canvas.
   void paintChildWithClipPath(RenderObject child, Point childPosition, Rect bounds, Path clipPath) {
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: ditto. Also, document bounds.

File: sky/packages/sky/lib/rendering/object.dart [hunk 7]
```diff
@@ -188,6 +234,11 @@ class PaintingContext {
...
   }
 
+  /// Paint a child with an opacity
+  ///
+  /// If the child needs compositing, the blending operation will be applied by
+  /// a compositing layer. Otherwise, the blending operation will be applied by
+  /// the canvas.
   void paintChildWithOpacity(RenderObject child,
                              Point childPosition,
                              Rect bounds,
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: document bounds

File: sky/packages/sky/lib/rendering/object.dart [hunk 8]
```diff
@@ -215,6 +266,11 @@ class PaintingContext {
       ..isAntiAlias = false;
   }
 
+  /// Paint a child with a color filter
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Since the arguments don't include a color filter, document how the color filter is built from the arguments.
Also document bounds.

File: sky/packages/sky/lib/rendering/object.dart [hunk 9]
```diff
@@ -238,12 +294,16 @@ class PaintingContext {
     }
   }
 
-  // do not call directly
+  /// Do not call directly
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Document what it does and the implications of calling it.

File: sky/packages/sky/lib/rendering/object.dart [hunk 10]
```diff
@@ -238,12 +294,16 @@ class PaintingContext {
...
+  ///
+  /// This function is visible so that it can be overridden in tests.
   void insertChild(RenderObject child, Offset offset) {
     child._paintWithContext(this, offset);
   }
 
-  // do not call directly
+  /// Do not call directly
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: ditto

File: sky/packages/sky/lib/rendering/object.dart [hunk 11]
```diff
@@ -285,8 +345,14 @@ class PaintingContext {
...
 }
 
+/// An abstract set of layout constraints
+///
+/// Concrete layout models (such as box) will create concrete subclasses to
+/// communicate layout constraints between parents and children.
 abstract class Constraints {
   const Constraints();
+
+  /// Whether there is exactly one layout possible given these constraints
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: One size, not one layout.

File: sky/packages/sky/lib/rendering/object.dart [hunk 12]
```diff
@@ -294,31 +360,45 @@ typedef void RenderObjectVisitor(RenderObject child);
...
 
-  // parentData is only for use by the RenderObject that actually lays this
-  // node out, and any other nodes who happen to know exactly what
-  // kind of node that is.
+  /// Data for use by the parent render object
+  ///
+  /// The parent data is used by the render object that lays out this object
+  /// (typically this object's parent in the render tree) to store information
+  /// relevant to itself and to any other nodes who happen to know exactly what
+  /// the data means. The parent data is opaque to the child.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Add the following information:
 - that the parentData object must inherit from ParentData (despite being typed as `dynamic`)
 - that the parentData field must not be directly set, except by calling setupParentData() on the parent node
 - that the parentData can be set before the child is added to the parent, by calling setupParentData on the future parent node
 - that the parent has responsibility over this field
 - that whether the child is allowed to access this field depends entirely on the convention defined by the ParentData subclass. (For example, BoxParentData is defined as being opaque to the child, so subclasses of RenderBox must not access their own parentData field.)

File: sky/packages/sky/lib/rendering/object.dart [hunk 13]
```diff
@@ -294,31 +360,45 @@ typedef void RenderObjectVisitor(RenderObject child);
...
   dynamic parentData; // TODO(ianh): change the type of this back to ParentData once the analyzer is cleverer
+
+  /// Override to setup parent data correctly for your children
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Also describe the conditions under which you could call this directly.

File: sky/packages/sky/lib/rendering/object.dart [hunk 14]
```diff
@@ -294,31 +360,45 @@ typedef void RenderObjectVisitor(RenderObject child);
...
     assert(debugCanPerformMutations);
     if (child.parentData is! ParentData)
       child.parentData = new ParentData();
   }
 
-  void adoptChild(RenderObject child) { // only for use by subclasses
-    // call this whenever you decide a node is a child
+  /// Called by subclases when they decide a render object is a child
+  ///
+  /// Only for use by subclasses
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Only for use by subclasses when changing their child lists; calling this in other cases will lead to an inconsistent tree and probably cause crashes.

File: sky/packages/sky/lib/rendering/object.dart [hunk 15]
```diff
@@ -294,31 +360,45 @@ typedef void RenderObjectVisitor(RenderObject child);
...
     super.adoptChild(child);
     markNeedsLayout();
-    markNeedsCompositingBitsUpdate();
+    _markNeedsCompositingBitsUpdate();
   }
-  void dropChild(RenderObject child) { // only for use by subclasses
+
+  /// Called by subclases when they decide a render object is no longer a child
+  ///
+  /// Only for use by subclasses
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: ditto

File: sky/packages/sky/lib/rendering/object.dart [hunk 16]
```diff
@@ -326,12 +406,13 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
     child.parentData.detach();
     super.dropChild(child);
     markNeedsLayout();
-    markNeedsCompositingBitsUpdate();
+    _markNeedsCompositingBitsUpdate();
   }
 
-  // Override in subclasses with children and call the visitor for each child.
+  /// Override in subclasses with children and call the visitor for each child
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: also describe how to use this method

File: sky/packages/sky/lib/rendering/object.dart [hunk 17]
```diff
@@ -326,12 +406,13 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
   void visitChildren(RenderObjectVisitor visitor) { }
 
+  /// Stored data useful for debugging exceptions
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: why document this one but not the other debug* fields?

File: sky/packages/sky/lib/rendering/object.dart [hunk 18]
```diff
@@ -376,10 +457,12 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
   static List<RenderObject> _nodesNeedingLayout = new List<RenderObject>();
   bool _needsLayout = true;
+  /// Whether this render object's layout information is dirty
   bool get needsLayout => _needsLayout;
   RenderObject _relayoutSubtreeRoot;
   bool _doingThisLayoutWithCallback = false;
   Constraints _constraints;
+  /// The layout constraints most recently supplied by the parent
   Constraints get constraints => _constraints;
   bool debugDoesMeetConstraints(); // override this in a subclass to verify that your state matches the constraints object
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: document this one, since you have to override it if you implement a new RenderObject subclass

File: sky/packages/sky/lib/rendering/object.dart [hunk 19]
```diff
@@ -396,6 +479,16 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+
+  /// Mark this render object's layout information as dirty
+  ///
+  /// Rather than eagerly updating layout information in response to writes into
+  /// this render object, we instead mark the layout information as dirty, which
+  /// schedules a visual update. As part of the visual update, the rendering
+  /// pipeline will update this render object's layout information.
+  ///
+  /// This mechanism batches the layout work so that multiple sequential writes
+  /// are coalesced, removing redundant computation.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Also document how this impacts needsLayout.
Also document how this decides to go mark a parent as needsLayout.

File: sky/packages/sky/lib/rendering/object.dart [hunk 20]
```diff
@@ -439,6 +538,12 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
     });
     _nodesNeedingLayout.add(this);
   }
+
+  /// Update the layout information for all dirty render objects
+  ///
+  /// This function is one of the core stages of the rendering pipeline. Layout
+  /// information is cleaned prior to painting so that render objects will
+  /// appear on screen in their up-to-date locations.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Give a pointer to the code that calls this, for people trying to track down how it all works.

File: sky/packages/sky/lib/rendering/object.dart [hunk 21]
```diff
@@ -427,6 +521,11 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
       });
     }
   }
+
+  /// Bootstrap the rendering pipeline by scheduling the very first layout
+  ///
+  /// Requires this render object to be attached and that this render object
+  /// is the root of the render tree.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Give a pointer to the code that calls this, for people trying to track down how it all works.

File: sky/packages/sky/lib/rendering/object.dart [hunk 22]
```diff
@@ -485,6 +590,28 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  /// This function is the main entry point for parents to ask their children to
+  /// update their layout information. The parent passes a constraints object,
+  /// which informs the child as which layouts are permissible. The child is
+  /// required to obey the given constraints.
+  ///
+  /// If the parent reads information computed during the child's layout, the
+  /// parent must pass true for parentUsesSize. In that case, the parent will be
+  /// marked as needing layout whenever the child is marked as needing layout
+  /// because the parent's layout information depends on the child's layout
+  /// information. If the parent passes false for parentUsesSize, the child can
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: This will encourage people to say `parentUsesSize: false` instead of just not mentioning it. Maybe rephrase that to encourage using the default.

File: sky/packages/sky/lib/rendering/object.dart [hunk 23]
```diff
@@ -485,6 +590,28 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  /// change its layout information (subject to the given constraints) without
+  /// informing the parent.
+  ///
+  /// Subclasses should not override layout directly. Instead, they should
+  /// override performResize and/or performLayout.
+  ///
+  /// Parent should call layout unconditionally. It is layout's responsibility
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Make that: The parent's performLayout method should call the layout of all its children unconditionally.

Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: and "It is the layout() method's responsibility (as implemented here) to return early..."

File: sky/packages/sky/lib/rendering/object.dart [hunk 24]
```diff
@@ -533,20 +660,39 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
     markNeedsPaint();
     assert(parent == this.parent); // TODO(ianh): Remove this once the analyzer is cleverer
   }
-  bool get sizedByParent => false; // return true if the constraints are the only input to the sizing algorithm (in particular, child nodes have no impact)
-  void performResize(); // set the local dimensions, using only the constraints (only called if sizedByParent is true)
+
+  /// Whether the constraints are the only input to the sizing algorithm (in particular, child nodes have no impact)
+  ///
+  /// Returning false is always correct, but returning true can be more
+  /// efficient in some siutations.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Correct the spelling of situations and elaborate on what some of those might be...

File: sky/packages/sky/lib/rendering/object.dart [hunk 25]
```diff
@@ -533,20 +660,39 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  bool get sizedByParent => false;
+
+  /// Updates the render objects size using only the constraints
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Mention that the subclasses should override this (if they set sizedByParent to true).

Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Add "Instead of calling this, call layout()." or some such.

File: sky/packages/sky/lib/rendering/object.dart [hunk 26]
```diff
@@ -533,20 +660,39 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  ///
+  /// This function is called only if [sizedByParent] is true.
+  void performResize();
+
+  /// Do the work of computing the layout for this render object
+  ///
+  /// Do not call this function directly. This function is called by layout when
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: "Do not call this function directly; call `layout()` instead."

File: sky/packages/sky/lib/rendering/object.dart [hunk 27]
```diff
@@ -533,20 +660,39 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  /// the dimensions of this render object. Instead, that work should be done by
+  /// [performResize]. If [sizedByParent] is false, then this function should
+  /// both change the dimensions of this render object and instruct its children
+  /// to layout.
+  ///
+  /// When calling [layout] on your children, pass true for parentUsesSize if
+  /// your layout information is dependent on your child's layout information.
+  /// Passing true for parentUsesSize ensures that this render object will
+  /// undergo layout if the child undergoes layout. Otherwise, the child can
+  /// changes its layout information without informing this render object.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Mention that you must call layout() on _all_ your children.

File: sky/packages/sky/lib/rendering/object.dart [hunk 28]
```diff
@@ -533,20 +660,39 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
-    // This is called during layout. If sizedByParent is true, then
-    // performLayout() should not change your dimensions, only do that
-    // in performResize(). If sizedByParent is false, then set both
-    // your dimensions and do your children's layout here.
-    //
-    // When calling layout() on your children, pass in
-    // "parentUsesSize: true" if your size or layout is dependent on
-    // your child's size or intrinsic dimensions.
+
+  /// Updates state to account for re-entrant layout and invokes callback
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Technically the layout isn't re-entrant with this.

This just allow you to mutate your child list (and the rest of that subtree) during performLayout().

File: sky/packages/sky/lib/rendering/object.dart [hunk 29]
```diff
@@ -590,36 +737,52 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
-  // same (in case ours changed). adoptChild() and dropChild() thus
-  // call markNeedsCompositingBitsUpdate().
   bool _needsCompositingBitsUpdate = true;
-  void markNeedsCompositingBitsUpdate() {
+  /// Mark the compositing state for this render object as dirty
+  ///
+  /// When the subtree is mutated, we need to recompute our [needsCompositing]
+  /// bit, and our ancestors need to do the same (in case ours changed).
+  /// Therefore, [adoptChild] and [dropChild] thus
+  /// call markNeedsCompositingBitsUpdate().
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: remove one of therefore and thus.

File: sky/packages/sky/lib/rendering/object.dart [hunk 30]
```diff
@@ -590,36 +737,52 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
 
   static List<RenderObject> _nodesNeedingPaint = new List<RenderObject>();
 
-  // Override this in subclasses to indicate that instances of your
-  // class need to have their own Layer. For example, videos.
+  /// Whether this render object paints using a composited layer
+  ///
+  /// Override this in subclasses to indicate that instances of your class need
+  /// to have their own compositing layer. For example, videos should return
+  /// true if they use hardware decoders.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: Mention that this must not change value across the lifetime of the object.

File: sky/packages/sky/lib/rendering/object.dart [hunk 31]
```diff
@@ -736,18 +922,50 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  /// When painting one of your children (via a paint child function on the
+  /// given context), the current canvas held by the context might change
+  /// because draw operations before and after painting children might need to
+  /// be recorded on separate compositing layers.
   void paint(PaintingContext context, Offset offset) { }
 
+  /// If this render object applies a transform before painting, apply that
+  /// transform to the given matrix
+  ///
+  /// Used by coodinate conversion functions to translate coordiantes local to
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: coordiantes typo

File: sky/packages/sky/lib/rendering/object.dart [hunk 32]
```diff
@@ -736,18 +922,50 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
+  /// Subclasses should override this function to provide a visual appearance
+  /// for themselves. The render object's local coordinate system is
+  /// axis-aligned with the coordinate system of the context's canvas and the
+  /// render object's local origin (i.e, x=0 and y=0) is placed at the given
+  /// offset in the context's canvas.
+  ///
+  /// Do not call this function directly. If you wish to paint yourself, call
+  /// [markNeedsPaint] instead to schedule a call to this function. If you wish
+  /// to paint one of your children, call one of the paint child functions on
+  /// the given context.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: add `, such as paintChild() or paintChildWithClipRect()` (or whatever they are called)

File: sky/packages/sky/lib/rendering/object.dart [hunk 33]
```diff
@@ -736,18 +922,50 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
...
 
 
   // EVENTS
 
+  /// Override this function to handle events that hit this render object
   EventDisposition handleEvent(sky.Event event, HitTestEntry entry) {
-    // override this if you have a client, to hand it to the client
-    // override this if you want to do anything with the event
     return EventDisposition.ignored;
   }
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: We should surface the HIT TESTING comment below too, somehow...

File: sky/packages/sky/lib/rendering/object.dart [hunk 34]
```diff
@@ -811,10 +1033,12 @@ double clamp({ double min: 0.0, double value: 0.0, double max: double.INFINITY }
 }
 
 
-// GENERIC MIXIN FOR RENDER NODES WITH ONE CHILD
-
+/// Generic mixin for render objects with one child
+///
+/// Provides a child model for a render object subclass that has a unique child
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: describe how to use this

File: sky/packages/sky/lib/rendering/object.dart [hunk 35]
```diff
@@ -842,12 +1066,13 @@ abstract class RenderObjectWithChildMixin<ChildType extends RenderObject> implem
...
-// GENERIC MIXIN FOR RENDER NODES WITH A LIST OF CHILDREN
-
+/// Parent data to support a doubly-linked list of children
 abstract class ContainerParentDataMixin<ChildType extends RenderObject> {
+  /// The previous sibling in the parent's child list
   ChildType previousSibling;
+  /// The next sibling in the parent's child list
   ChildType nextSibling;
+
   void detachSiblings() {
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: /// Clear the sibling pointers.

File: sky/packages/sky/lib/rendering/object.dart [hunk 36]
```diff
@@ -866,6 +1091,10 @@ abstract class ContainerParentDataMixin<ChildType extends RenderObject> {
   }
 }
 
+/// Generic mixin for render objects with a list of children
+///
+/// Provides a child model for a render object subclass that has a doubly-linked
+/// list of children.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: document how to use this too (especially the part about providing a ParentDataType...)

File: sky/packages/sky/lib/rendering/object.dart [hunk 37]
```diff
@@ -934,6 +1164,9 @@ abstract class ContainerRenderObjectMixin<ChildType extends RenderObject, Parent
       }
     }
   }
+  /// Insert child into this render object's child list before the given child
+  ///
+  /// To insert a child at the end of the child list, pass null for before.
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: (or omit `before`)

File: sky/packages/sky/lib/rendering/object.dart [hunk 38]
```diff
@@ -971,10 +1207,18 @@ abstract class ContainerRenderObjectMixin<ChildType extends RenderObject, Parent
...
+
+  /// Remove this child from the child list
+  ///
+  /// Requires the child to be present in the child list.
   void remove(ChildType child) {
     _removeFromChildList(child);
     dropChild(child);
   }
+
+  /// Remove all thie children from this render object's child list
```
Comment (Hixie) [inline sky/packages/sky/lib/rendering/object.dart]: thie typo

**Comment by Hixie** [comment]
LGTM... modulo comments. :-)
